src/beanmachine/paic2/src/MLIRBuilder.cpp

```diff
@@ -0,0 +1,413 @@
+                return builder.getNoneType();
+            default:
+                throw std::invalid_argument("unrecognized primitive type");
+        }
+    }
+
+    // There are two types supported: PrimitiveType and the builtin type WorldType
+    // A World type is compiled into a composite type that contains an array, sets of function pointers, and metadata
+    mlir::Type getType(std::shared_ptr<paic2::Type> type) {
+        if(auto *var = dyn_cast<paic2::PrimitiveType>(type.get())){
+            return typeForCode(var->code());
+        } else if (auto *var = dyn_cast<paic2::WorldType>(type.get())) {
+            std::vector<mlir::Type> members;
+            mlir::Type elementType = typeForCode(var->nodeType());
+            ArrayRef<int64_t> shape(var->length());
+            auto dataType = mlir::RankedTensorType::get(shape, elementType);
+            members.push_back(dataType);
+            mlir::Type structType = mlir::bm::WorldType::get(members);
+            return structType;
+        }
+        throw std::invalid_argument("unrecognized primitive type");
+    }
+
+    mlir::LogicalResult declare(llvm::StringRef var, mlir::Value value) {
+        if (symbolTable.count(var))
+            return mlir::failure();
+        symbolTable.insert(var, value);
+        return mlir::success();
+    }
+
+    mlir::Value mlirGen(paic2::GetValNode* expr) {
+        if (auto variable = symbolTable.lookup(expr->getName()))
+            return variable;
+
+        emitError(loc(expr->loc()), "error: unknown variable '")
+                << expr->getName() << "'";
+        return nullptr;
+    }
+
+    mlir::LogicalResult mlirGen(paic2::ReturnNode* ret) {
+        auto location = loc(ret->loc());
+        mlir::Value expr = nullptr;
+        if (ret->getValue()) {
+            if (!(expr = mlirGen(ret->getValue().get())))
+                return mlir::failure();
+        }
+        builder.create<mlir::bm::ReturnOp>(location, expr ? makeArrayRef(expr) : ArrayRef<mlir::Value>());
+        return mlir::success();
+    }
+
+    mlir::Value mlirGen(paic2::CallNode* call) {
+        llvm::StringRef callee = call->getCallee();
+        auto location = loc(call->loc());
+
+        // Codegen the operands first.
+        SmallVector<mlir::Value, 4> operands;
+        for (std::shared_ptr<paic2::Expression> expr : call->getArgs()) {
+            auto arg = mlirGen(expr.get());
+            if (!arg)
+                return nullptr;
+            operands.push_back(arg);
+        }
+
+        // TODO: make map and remove magic strings
+        if (callee == "times") {
+            if (call->getArgs().size() != 2) {
+                emitError(location, "MLIR codegen encountered an error: times "
+                                    "accepts exactly two arguments");
+                return nullptr;
+            }
+            return builder.create<mlir::arith::MulFOp>(location, operands[0], operands[1]);
+        } else if (callee == "math.pow") {
```

Comment:
expose these as enums on a NumberLike enum in Paic2 ast so that clients are using types rather than strings to communicate builtin methods. This should be done before control flow is added since we need LT, GT, GEQ, and LEQ for control flow. So, it should be done in this PR I think